codetransformer/utils/immutable.py

```diff
@@ -178,26 +188,105 @@ def _wrapinit(init):
-
-        missing_slots = tuple(
-            filter(lambda s: not hasattr(self, s), self.__slots__),
+    try:
+        spec = getfullargspec(init)
+    except TypeError:
+        # we cannot preserve the type signature.
+        def __init__(*args, **kwargs):
+            self = args[0]
+            __setattr__._initializing.add(self)
+            init(*args, **kwargs)
+            __setattr__._initializing.remove(self)
+            _check_missing_slots(self)
+
+        return __init__
+
+    args = spec.args
+    varargs = spec.varargs
+    if not (args or varargs):
+        raise TypeError(
+            "%r must accept at least one positional argument for 'self'" %
+            getattr(init, '__qualname__', getattr(init, '__name__', init)),
         )
-        if missing_slots:
-            raise TypeError(
-                'not all slots initialized in __init__, missing: {0}'.format(
-                    missing_slots,
-                ),
-            )
 
+    if not args:
+        self = '%s[0]' % varargs
+        forward = argspec = '*' + varargs
+    else:
+        self = args[0]
+        forward = argspec = ', '.join(args)
+
+    if args and varargs:
+        forward = '%s, *%s' % (forward, spec.varargs)
+        argspec = '%s, *%s' % (argspec, spec.varargs)
+    if spec.kwonlyargs:
+        forward = '%s, %s' % (
+            forward,
+            ', '.join(map('{0}={0}'.format, spec.kwonlyargs))
+        )
+        argspec = '%s,%s%s' % (
+            argspec,
+            '*, ' if not spec.varargs else '',
+            ', '.join(spec.kwonlyargs),
+        )
+    if spec.varkw:
+        forward = '%s, **%s' % (forward, spec.varkw)
+        argspec = '%s, **%s' % (argspec, spec.varkw)
+
+    ns = {
+        '__init': init,
+        '__initializing': __setattr__._initializing,
+        '__check_missing_slots': _check_missing_slots,
+    }
+    exec(
+        dedent(
+            """\
+            def __init__({argspec}):
+                __initializing.add({self})
+                __init({forward})
+                __initializing.remove({self})
+                __check_missing_slots({self})
+            """.format(
+                argspec=argspec,
+                self=self,
+                forward=forward,
+            ),
+        ),
+        ns,
+    )
+    __init__ = ns['__init__']
+    __init__.__defaults__ = spec.defaults
+    __init__.__kwdefaults__ = spec.kwonlydefaults
+    __init__.__annotations__ = spec.annotations
     return __init__
 
 
+def _check_missing_slots(ob):
+    """Check that all slots have been initialized when a custom __init__ method
+    is provided.
+
+    Parameters
+    ----------
+    ob : immutable
+        The instance that was just initialized.
+
+    Raises
+    ------
+    TypeError
+        Raised when the instance has not set values that are named in the
+        __slots__.
+    """
+    missing_slots = tuple(
+        filter(lambda s: not hasattr(ob, s), ob.__slots__),
```

Comment:
just having a slot does not meant that the value has been initialized:

```python
In [1]: class C:
   ...:     __slots__ = 'a', 'b'
   ...:     

In [2]: c = C()

In [3]: c.a
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-3-045efd26f8b5> in <module>()
----> 1 c.a

AttributeError: a
```

This is called when the user has provided their own constructor to ensure that they are not leaving the object partially initialized.